Octane is here (#455)
ember-learn/ember-blog: source/2019-12-20-octane-is-here.md

---
title: Octane is Here
author: Yehuda Katz
tags: Recent Posts, 2019, Announcement

> responsive: true

Reviewer: ```suggestion
alias: 2019-12-20-octane-is-here.md
responsive: true
```

## Octane is More Fun

The Ember Octane edition is, first and foremost, about making it easier and more fun to build Ember applications.

The centerpiece of Octane's ergonomic improvements are two big changes to the core of Ember: a new component model and a new reactivity system.

For existing Ember users, both the new component model and the new reactivity system are fully opt-in and fully interoperable with existing code. Upgrading an Ember 3.14 app to Ember 3.15 is a compatible change, as the version number would suggest.

## Glimmer Components

The first big improvement in Ember Octane is Glimmer Components. Ember has had a single component system since Ember 1.0, based on JavaScript syntax that was available at the time.

#### Before: Classic Components

The thing that jumps out at you when you look at classic components is that you configure a "root element" using a JavaScript microsyntax.

```js
export default EmberComponent.extend({
  tagName: 'p',
  classNames: ["tooltip"],
  classNameBindings: ["isEnabled:enabled", "isActive:active"],
})
```

#### After: Glimmer Components

> In contrast, Glimmer components allow you to treat the root element just like any other element. This substantially simplifies the component model, eliminating the special cases that come from having a second API just for working with the root element of a component.

Reviewer: ```suggestion
<!--alex ignore special-->
In contrast, Glimmer components allow you to treat the root element just like any other element. This substantially simplifies the component model, eliminating the special cases that come from having a second API just for working with the root element of a component.
```

## Getting Started

Whether you're a new Ember developer, coming back to Ember after many years, or an existing Ember developer, the fastest and easiest way to get learn how to build applications the Octane way is to run through [the refreshed tutorial](https://guides.emberjs.com/release/tutorial/part-1/).

Once you got through the tutorial, it's time to build something real for fun. The Ember addon ecosystem is a great part of Ember, so you'll want to reach for addons to speed up the process of building your project.

[Ember Observer](https://emberobserver.com/) is a directory for the Ember addon ecosystem. Each addon gets a quality score based on a human review of formal criteria like the existence of a meaningful README, whether the addon has an automated build, and whether the addon is maintained by more than one person. This week, it will also indicate whether an addon is Octane Ready.

Because of the details of Octane's compatibility story, most addons should be Octane Ready without any changes. Ember Observer will help the community proactively identify and fix Octane problems in maintained packages.

## A Closer Look at Seamless Interop

> In addition to the elimination of computed properties, the Glimmer reactivity model also doesn't include special Ember proxies or observers. The Octane reactivity model is more powerful than the classic one, but it's much easier to use.

Reviewer: ```suggestion
<!--alex ignore special-->
In addition to the elimination of computed properties, the Glimmer reactivity model also doesn't include special Ember proxies or observers. The Octane reactivity model is more powerful than the classic one, but it's much easier to use.
```

```js
import Mixin from '@ember/object/mixin';

export default Mixin.create({
  didInsertElement() {
    this._super();
    activateTabs(this.element);
  }

  willDestroyElement() {
    this._super();
    deactivateTabs(this.element);
  }
})
```

And then you would use it in a component like this:

```js
import Component from '@ember/component';

export default EmberComponent.extend(Tabs, {
  // ...
});
```

The drawbacks of using mixins for UI composition are well-described [across](https://reactjs.org/blog/2016/07/13/mixins-considered-harmful.html) [the](https://blog.logrocket.com/hooks-are-coming-to-vue/) [JavaScript](https://github.com/sveltejs/svelte/issues/1041#issuecomment-353660981) [ecosystem](https://github.com/emberjs/rfcs/issues/534). The most glaring issue is naming conflicts. Any method on a mixin might conflict with a method on any other mixin, with no good way to resolve the conflicts.

In the context of Ember, there's another issue with using Ember Component mixins for reusable DOM behavior. If you want to use the `Tabs` mixin on an element, you need to turn that element into a component with a JavaScript class, which is pretty awkward.

> You can still use mixins in Ember 3.15, and addons may still provide mixins for you to use.

Reviewer: ```suggestion
While we do recommend you avoid mixins, you can still use them in Ember 3.15. Addons may also still provide mixins for you to use.
```
